### PR TITLE
Internet Gateway terminal codes

### PR DESCRIPTION
**Pull request by LikithaVemulapalli**
Issue #, if available: [1362](https://github.com/aws-controllers-k8s/community/issues/1362)

Description of changes:

Made the following changes in order to map exceptions for **internet gateway** resource in ec2-controller.
      - Added terminal codes in the "generator.yaml" for internet gateway resource.
      - Added few integration tests for internet gateway resource that cover invalid and malformed vpc-id.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by brycahta — COMMENTED** [review]
/ok-to-test

File: generator.yaml [hunk 1]
```diff
@@ -233,6 +233,15 @@ resources:
       sdk_read_many_post_build_request:
         template_path: hooks/elastic_ip_address/sdk_read_many_post_build_request.go.tpl
   InternetGateway:
+    exceptions:
+      terminal_codes:
+      - InvalidInternetGatewayID.NotFound
+      - InvalidInternetGatewayId.Malformed
+      - Resource.AlreadyAssociated
```
Comment (brycahta) [inline generator.yaml:240]: I don't think exceptions such as `Resource.AlreadyAssociated` and `Gateway.NotAttached` are possible given the [attach/detach implementation](https://github.com/aws-controllers-k8s/ec2-controller/blob/main/pkg/resource/internet_gateway/hook.go#L39). Am I missing context or can we remove?

File: test/e2e/tests/test_internet_gateway.py [hunk 2]
```diff
@@ -181,4 +181,67 @@ def test_vpc_association(self, ec2_client, empty_vpc):
         time.sleep(DELETE_WAIT_AFTER_SECONDS)
 
         # Check Internet Gateway no longer exists in AWS
-        ec2_validator.assert_internet_gateway(resource_id, exists=False)
\ No newline at end of file
+        ec2_validator.assert_internet_gateway(resource_id, exists=False)
+
+    def test_terminal_condition_invalid_vpcid(self):
+        test_resource_values = REPLACEMENT_VALUES.copy()
+        resource_name = random_suffix_name("ig-invalid-vpcid", 24)
```
Comment (brycahta) [inline test/e2e/tests/test_internet_gateway.py:188]: `resource_name` is being assigned a `vpcid` value in both tests. Can you update so the name reflects the *internet gateway* resource?

**Review by brycahta — CHANGES_REQUESTED** [review]
Nice first PR! 🎉 

Had a couple comments on minor things, but overall very clean

File: generator.yaml (hunk 1 above)
Comment (LikithaVemulapalli) [reply generator.yaml:240]: Yes I agree with the statement that these exceptions are not possible considering our implementation. We cannot test them, thought that we can include them in terminal codes, I can remove them!

File: test/e2e/tests/test_internet_gateway.py (hunk 2 above)
Comment (LikithaVemulapalli) [reply test/e2e/tests/test_internet_gateway.py:188]: I added resource_name values as per the failure scenarios, will update with the names that reflects the internet gateway resource.

File: generator.yaml (hunk 1 above)
Comment (brycahta) [reply generator.yaml]: I'm thinking we just include the exceptions that are possible/testable:
```
      - InvalidVpcId.Malformed
      - InvalidVpcID.NotFound
```

@RedbackThomson @jaypipes  do y'all have other thoughts/opinions?

Comment (vijtrip2) [reply generator.yaml]: My 2 cents:

TL;DR: Vet carefully the exceptions being marked as terminal because missing an exception as terminal is better than marking a recoverable exception as terminal.

---
 
Terminal exceptions cause the reconciliation to stop. The exceptions added to the terminal exception should be vetted carefully. 
In case of some terminal exceptions not being added to this block, controller will just retry with exponential backoff. However if some recoverable exception gets added here, the reconciliation will cease.

---

Also NotFound errors should not be marked as terminal because we have seen customer submit their manifests out of order and some resources being created after their dependent resource.
In resource reference we handle this scenario, but have seen customer complaints in the past(they weren't using references)

Comment (brycahta) [reply generator.yaml]: Appreciate the insights and feedback @vijtrip2 -- very helpful 👍 

All your points make sense to me. Basically, we only want `InvalidVpcId.Malformed` as a terminal exception for `InternetGateway` (Likitha & I have deep dived the other exceptions already) because reconciliation cannot "correct" this to a valid state. We can always add more exceptions to be terminal later should any issues arise. Feel free to correct anything/lmk if that sounds good. Thanks again

**Review by brycahta — COMMENTED** [review]
/lgtm
